src/main/kotlin/org/wfanet/measurement/dataprovider/common/GrpcElGamalPublicKeyCache.kt

```diff
@@ -0,0 +1,41 @@
+// Copyright 2021 The Cross-Media Measurement Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.wfanet.measurement.dataprovider.common
+
+import org.wfanet.measurement.api.v2alpha.CombinedPublicKey
+import org.wfanet.measurement.api.v2alpha.CombinedPublicKeysGrpcKt.CombinedPublicKeysCoroutineStub
+import org.wfanet.measurement.api.v2alpha.ElGamalPublicKey
+import org.wfanet.measurement.api.v2alpha.GetCombinedPublicKeyRequest
+
+/**
+ * Implementation of [ElGamalPublicKey] that caches results obtained from gRPC calls.
+ */
+class GrpcElGamalPublicKeyCache(
+  private val combinedPublicKeys: CombinedPublicKeysCoroutineStub
+) : ElGamalPublicKeyStore {
+  private val cache = mutableMapOf<CombinedPublicKey.Key, ElGamalPublicKey>()
+
+  override suspend fun get(key: CombinedPublicKey.Key): ElGamalPublicKey {
+    return cache.getOrPut(key) { getFromGrpc(key) }
+  }
+
+  private suspend fun getFromGrpc(key: CombinedPublicKey.Key): ElGamalPublicKey {
```

Comment:
what is this Grpc for? My understanding from various conversions is that the individual duchies' ElGamalKeys are embedded in the requisition. The EDP would combine them locally and use the result to do the encryption.

Since each requisition uses a new set of keys, it seems that caching the keys is not necessary.